File: framework/cei_java/src/main/java/cn/ma/cei/sdk/impl/RestfulRequest.java
package cn.ma.cei.sdk.impl;

import cn.ma.cei.sdk.exception.CEIException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import javafx.util.Pair;

import java.util.LinkedList;
import java.util.List;

public class RestfulRequest {

    private String url;
    private String target;
    private Method method;
    private RestfulOption options;
    private byte[] requestBody_ = null;
    private List<Pair<String, String>> queryString_ = new LinkedList<>();
    private List<Pair<String, String>> headers_ = new LinkedList<>();

    public RestfulRequest(RestfulOption options) {
        this.options = options;
    }

    public String getUrl() {
        return url;
    }

    public String getTarget() {
        return target;
    }

    public Method getMethod() {
        return method;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public void setMethod(Method method) {
        this.method = method;
    }

    public void setTarget(String target) {
        this.target = target;
    }

    public void addHeader(String name, String value) {
        headers_.add(new Pair<>(name, value));
    }

    public void addQueryString(String name, String value) {
        if (value == null) {
            return;
        }
        queryString_.add(new Pair<>(name, value));
    }

    public void addQueryString(String name, Boolean value) {
        addQueryString(name, value.toString());
    }

    public String buildQueryString() {
        StringBuilder builder = new StringBuilder();
        queryString_.forEach((item) -> {
            if (!("").equals(builder.toString())) {
                builder.append('&');
            } else {
                builder.append('?');
            }
            builder.append(item.getKey());
            builder.append("=");
            builder.append(urlEncode(item.getValue()));
        });
        return builder.toString();
    }

    public List<Pair<String, String>> getHeaders() {
        return headers_;
    }

    public void setRequestBody(byte[] body) {
        requestBody_ = body;
    }

    private static String urlEncode(String s) {
        try {
            return URLEncoder.encode(s, "UTF-8").replaceAll("\\+", "%20");
        } catch (UnsupportedEncodingException e) {
            throw new CEIException("[URL] UTF-8 encoding not supported!");
        }
    }

    public enum Method {
        GET,
        POST,
        DELETE
    }
}

File: framework/cei_golang/src/impl/restful/restfulrequest.go
package restful

import (
	"strconv"
)

type Method string

const (
	GET  Method = "GET"
	POST Method = "POST"
)

type RestfulRequest struct {
	url          string
	target       string
	method       Method
	headers      []string
	queryStrings []string
}

func NewRestfulRequest(options *RestfulOptions) *RestfulRequest {
	instance := new(RestfulRequest)
	return instance
}

func (inst *RestfulRequest) SetTarget(target string) {
	inst.target = target
}

func (inst *RestfulRequest) GetTarget() string {
	return inst.target
}

func (inst *RestfulRequest) SetMethod(method Method) {
	inst.method = method
}

func (inst *RestfulRequest) GetMethod() Method {
	return inst.method
}

func (inst *RestfulRequest) AddHeader(key string, value interface{}) {
	inst.headers = append(inst.headers, key, toString(value))
}

func (inst *RestfulRequest) AddQueryString(key string, value interface{}) {
	inst.queryStrings = append(inst.queryStrings, key, toString(value))
}

func toString(value interface{}) string {
	switch v := value.(type) {
	case string:
		return string(v)
	case bool:
		return strconv.FormatBool(v)
	case uint8:
		return strconv.FormatUint(uint64(v), 10)
	case uint16:
		return strconv.FormatUint(uint64(v), 10)
	case uint32:
		return strconv.FormatUint(uint64(v), 10)
	case uint64:
		return strconv.FormatUint(uint64(v), 10)
	case uint:
		return strconv.FormatUint(uint64(v), 10)
	case int8:
		return strconv.FormatInt(int64(v), 10)
	case int16:
		return strconv.FormatInt(int64(v), 10)
	case int32:
		return strconv.FormatInt(int64(v), 10)
	case int64:
		return strconv.FormatInt(int64(v), 10)
	case int:
		return strconv.FormatInt(int64(v), 10)
	case float32:
		return strconv.FormatFloat(float64(v), 'f', -1, 64)
	case float64:
		return strconv.FormatFloat(float64(v), 'f', -1, 64)
	default:
		// TODO not supported
		return ""
	}
}

func (inst *RestfulRequest) GetHeaders() []string {
	return inst.headers
}

func (inst *RestfulRequest) SetURL(value string) {
	inst.url = value
}

func (inst *RestfulRequest) GetURL() string {
	return inst.url
}

func (inst *RestfulRequest) BuildQueryString() string {
	result := ""
	for index := 0; index < len(inst.queryStrings); index += 2 {
		if len(result) != 0 {
			result += "&"
		}
		result += inst.queryStrings[index]
		result += "="
		result += inst.queryStrings[index+1]
	}
	return "?" + result
}
